README.md
# Ruff2BitBucket

[ruff](https://docs.astral.sh/ruff/) stands out as an exceptionally swift Python linter and code formatter, written in Rust. This tool not only expedites the process of identifying and correcting code issues but also demonstrates a powerful synergy between the efficiency of Rust and the versatility of Python, providing developers with a robust solution for enhancing code quality and adherence to coding standards.

[BitBucket](https://bitbucket.org/), an Atlassian product, serves as a comprehensive platform offering a visual Git interface, essentially presenting developers with an intuitive and user-friendly environment for managing their version-controlled code repositories.

Utilizing the ruff2bitbucket tool facilitates the seamless execution of your ruff validations, enabling you to effortlessly execute these checks and subsequently upload the results as a cohesive code insight directly into your BitBucket repository.

## Installation
```shell
pip install ruff2bitbucket
```

## Usage
Change your directory to where the configuration file is stored and run:
```shell
ruff2bitbucket [--user ...] [--pass ...]
```

⚠ It's important that you are in the right directory!

## Configuration
### ruff
For a comprehensive understanding of how to enhance and customize your code analysis, it is highly recommended to consult the [ruff configuration docs](https://docs.astral.sh/ruff/configuration/) and the [ruff settings docs](https://docs.astral.sh/ruff/settings/) where detailed instructions are provided on the modification of your pyproject.toml (or ruff.toml, or .ruff.toml) file, allowing you to tailor these configuration files to incorporate a broader range of checks and ensure a more thorough examination of your codebase.

### BitBucket
To begin configuring code insights, ensure that you have the necessary permissions in your Bitbucket repository.
- Navigate to your repository and click on "Settings" in the left sidebar.
- From the settings menu, select "Code insights" and then choose "Repository settings".
- Here, you can enable various code insights features.
- Add "ruff2bitbucket" as a code insight to enable this in your Pull Requests.

- By following these steps, you'll have the `ruff2bitbucket` code insight configured in Bitbucket, streamline the code review process. This integration fosters a more collaborative and efficient development environment, ultimately leading to higher-quality software.

## Security
When neither a username nor a password is explicitly provided (via the command line), the system will conduct a scan of environment variables for potential matches resembling `USER`/`USR` and `PASS`/`PW`.
Upon identifying a match, these discovered credentials will be employed to seamlessly attach annotations to the current commit hash, ensuring a streamlined process for handling authentication in the given environment.
This approach enhances flexibility while maintaining a focus on security, allowing users to conveniently utilize environment variables for authentication without compromising the integrity of the process.

CHANGELOG.md
# 0.3.0
* Reducing python minimal version to 3.7. This is the lowest version that ruff can be installed with.

# 0.2.0
* Adding `__version__`
* Adjusted logger levels
* Bugfix for `distutils.spawn.has_executable` which is deprectated in py3.10 in favor of `shutil.which`
* Bugfix for error when the insight triggered an error on BitBucket
* Bugfix for wrong variable name causing the program to crash during upload of a code insight.

# 0.1.0
* First version supporting basic ruff validation of a repository.


tests/conftest.py
import logging
import os
import subprocess
import sys

import pytest
from pytest_mock import MockerFixture
from ruff2bitbucket.bitbucket import get_repo_info
from ruff2bitbucket.credentials import get_credentials
from ruff2bitbucket.git import get_current_git_commit_hash, get_current_repo_uri


@pytest.fixture(autouse=True)
def _credentials_environment_setup(monkeypatch: pytest.MonkeyPatch) -> None:
    """Ensure that no leakage from the build environment happens to the tests."""

    monkeypatch.setattr(sys, "argv", ["script"])
    monkeypatch.setattr(os, "environ", {})


@pytest.fixture(autouse=True)
def _inject_run_commands(mocker: MockerFixture) -> None:
    def local_run(*cmd: str, check: bool) -> subprocess.CompletedProcess:
        if cmd == ("git", "config", "--get", "remote.origin.url"):
            assert check is True
            return subprocess.CompletedProcess("", 0, "https://localhost:12345/scm/abc/repository.git\n\n")

        if cmd == ("git", "rev-parse", "HEAD"):
            assert check is True
            return subprocess.CompletedProcess("", 0, "abcde_commit_hash_fghij\n\n")

        if cmd == ("ruff", "check", "--no-fix", "."):
            assert check is False
            return subprocess.CompletedProcess(
                "",
                0,
                (
                    "src/some_repo/filter/fltr.py:337:21: G004 Logging statement uses f-string\n"
                    "src/some_repo/filter/wrk.py:24:66: Q000 [*] Single quotes found but double quotes preferred\n"
                    "[..]\n"
                    "Found 1592 errors.\n"
                    "[*] 993 fixable with the `--fix` option (218 hidden fixes can be enabled with the `--unsafe-fixes` option).\n"  # noqa: E501
                ),
            )

        if cmd == ("ruff", "format", "--check", "."):
            assert check is False
            return subprocess.CompletedProcess(
                "",
                0,
                (
                    "Would reformat: src/some_repo/filter/fltr.py\n"
                    "Would reformat: src/some_repo/filter/wrk.py\n"
                    "37 files would be reformatted, 9 files left unchanged\n"
                ),
            )

        raise ValueError(f"Unknown call to: {cmd}")

    mocker.patch("ruff2bitbucket.git.run", new=local_run)
    mocker.patch("ruff2bitbucket.capturer.run", new=local_run)


@pytest.fixture(autouse=True)
def _enable_all_executables(mocker: MockerFixture) -> None:
    mocker.patch("shutil.which", return_value="I'm there!")


@pytest.fixture(autouse=True)
def _cleanup_caches() -> None:
    yield

    get_credentials.cache_clear()
    get_repo_info.cache_clear()
    get_current_git_commit_hash.cache_clear()
    get_current_repo_uri.cache_clear()


@pytest.fixture(autouse=True)
def _cleanup_get_repo_info_cache() -> None:
    return


@pytest.fixture()
def caplog(mocker: MockerFixture, caplog: pytest.LogCaptureFixture) -> pytest.LogCaptureFixture:
    root = logging.getLogger()

    mocker.patch.object(root, "disabled", new=False)
    mocker.patch.object(root, "handlers", new=[])
    mocker.patch.object(root, "level", new=logging.NOTSET)

    return caplog


tests/test_git.py
# Not much to test in here...
from subprocess import CompletedProcess

import pytest
from pytest_mock import MockerFixture
from pytest_mock.plugin import MockType
from ruff2bitbucket.git import get_current_git_commit_hash, get_current_repo_uri


@pytest.fixture()
def mock_run(mocker: MockerFixture) -> MockType:
    return mocker.patch("ruff2bitbucket.git.run", return_value=CompletedProcess("", 0, "abc", "def"))


def test_get_current_git_commit_hash(mock_run: MockType) -> None:
    assert get_current_git_commit_hash() == "abc"
    mock_run.assert_called_once_with("git", "rev-parse", "HEAD", check=True)

    mock_run.reset_mock()

    assert get_current_git_commit_hash() == "abc"
    mock_run.assert_not_called()


def test_get_current_repo_uri(mock_run: MockType) -> None:
    assert get_current_repo_uri() == "abc"
    mock_run.assert_called_once_with("git", "config", "--get", "remote.origin.url", check=True)

    mock_run.reset_mock()

    assert get_current_repo_uri() == "abc"
    mock_run.assert_not_called()


tests/test_main.py
import logging

import pytest
from pytest_mock import MockerFixture
from ruff2bitbucket import main

base_url = (
    "https://localhost:12345/rest/insights/latest/projects/abc/repos/"
    "repository/commits/abcde_commit_hash_fghij/reports/ruff2bitbucket"
)


@pytest.fixture(autouse=True)
def _setting_default_credential(monkeypatch: pytest.MonkeyPatch) -> None:
    monkeypatch.setenv("CRED_USER", "USER")
    monkeypatch.setenv("CRED_PASSWORD", "PASS")


def test_main_no_credentials(mocker: MockerFixture, caplog: pytest.LogCaptureFixture) -> None:
    mocker.patch("os.environ", new={})

    with pytest.raises(SystemExit) as ex:
        main()

    assert ex.value.code == 1
    assert len(caplog.records) == 1
    assert caplog.records[0].levelname == "ERROR"
    assert caplog.records[0].message == "No valid credentials found."


def test_main_no_valid_credentials_found(mocker: MockerFixture, caplog: pytest.LogCaptureFixture) -> None:
    put_mock = mocker.patch("requests.put")
    put_mock.return_value = mocker.Mock(status_code=401)  # UnAuthorized

    with pytest.raises(SystemExit) as ex:
        main()

    assert ex.value.code == 1
    assert len(caplog.records) == 1
    assert caplog.records[0].levelname == "ERROR"
    assert caplog.records[0].message == "Cannot upload the report to bitbucket. No valid user/pass found."


def test_main_some_bitbucket_error_happened_on_statistics(
    mocker: MockerFixture, caplog: pytest.LogCaptureFixture
) -> None:
    put_mock = mocker.patch("requests.put")
    put_mock.return_value.status_code = 400
    put_mock.return_value.json.return_value = {"dummy": 123}

    main()

    warning_records = [rec for rec in caplog.records if rec.levelname == "WARNING"]
    expected = [
        f"'PUT {base_url}' reported one or more errors:",
        "{",
        '    "dummy": 123',
        "}",
    ]

    assert len(warning_records) == len(expected)
    for idx, message in enumerate(expected):
        assert message == warning_records[idx].message


def test_main_some_bitbucket_error_happened_on_annotations(
    mocker: MockerFixture, caplog: pytest.LogCaptureFixture
) -> None:
    put_mock = mocker.patch("requests.put")
    put_mock.return_value.status_code = 404
    put_mock.return_value.json.return_value = {"dummy": 123}

    main()

    warning_records = [rec for rec in caplog.records if rec.levelname == "WARNING"]
    expected = [
        f"'PUT {base_url}/annotations' reported one or more errors:",
        "{",
        '    "dummy": 123',
        "}",
    ]

    assert len(warning_records) == len(expected)
    for idx, message in enumerate(expected):
        assert message == warning_records[idx].message


def test_main_no_errors_occurred(mocker: MockerFixture, caplog: pytest.LogCaptureFixture) -> None:
    mocker.patch("ruff2bitbucket.__main__.check_code_mistakes", return_value=[])
    mocker.patch("ruff2bitbucket.__main__.check_formatting", return_value=[])
    put_mock = mocker.patch("requests.put")
    caplog.set_level(logging.INFO)

    with pytest.raises(SystemExit) as ex:
        main()

    assert ex.value.code == 0

    put_mock.assert_not_called()

    assert len(caplog.records) == 1
    assert caplog.records[0].levelname == "INFO"
    assert caplog.records[0].message == "no errors detected. No report will be uploaded."


def test_main_happy_flow(mocker: MockerFixture) -> None:
    put_mock = mocker.patch("requests.put")
    put_mock.return_value = mocker.Mock(status_code=200)

    main()

    assert put_mock.call_count == 2

    put_mock.assert_any_call(
        base_url,
        json={
            "result": "FAIL",
            "title": "ruff report",
            "reporter": "ruff2bitbucket",
            "report_type": "CODE_SMELL",
            "data": [
                {"title": "Need reformat", "type": "NUMBER", "value": 2},
                {"title": "Issue count", "type": "NUMBER", "value": 2},
            ],
        },
        auth=("USER", "PASS"),
    )

    put_mock.assert_any_call(
        f"{base_url}/annotations",
        json={
            "annotations": [
                {
                    "reportKey": "ruff2bitbucket",
                    "path": "src/some_repo/filter/fltr.py",
                    "line": 337,
                    "message": "G004 Logging statement uses f-string",
                    "severity": "LOW",
                    "type": "CODE_SMELL",
                },
                {
                    "reportKey": "ruff2bitbucket",
                    "path": "src/some_repo/filter/wrk.py",
                    "line": 24,
                    "message": "Q000 [*] Single quotes found but double quotes preferred",
                    "severity": "LOW",
                    "type": "CODE_SMELL",
                },
                {
                    "reportKey": "ruff2bitbucket",
                    "path": "src/some_repo/filter/fltr.py",
                    "line": 0,
                    "message": "Would reformat",
                    "severity": "LOW",
                    "type": "CODE_SMELL",
                },
                {
                    "reportKey": "ruff2bitbucket",
                    "path": "src/some_repo/filter/wrk.py",
                    "line": 0,
                    "message": "Would reformat",
                    "severity": "LOW",
                    "type": "CODE_SMELL",
                },
            ]
        },
        auth=("USER", "PASS"),
    )


src/ruff2bitbucket/capturer.py
import contextlib
import logging
import re
import shutil
from typing import Iterable

from .common import CapturedLine, run

__all__ = ["check_code_mistakes", "check_formatting"]


capturer = re.compile(r"^(?P<filename>.*?):(?P<line>\d+):(?P<column>\d+):\s*(?P<description>.*)$", flags=re.IGNORECASE)
formater = re.compile(r"^(?P<description>Would reformat):?\s*(?P<filename>.*?)$", flags=re.IGNORECASE)
logger = logging.getLogger(__name__)


def has_executable(name: str) -> bool:
    return shutil.which(name) is not None


def yield_from_regex(*cmd: str, regex_to_use: re.Pattern) -> Iterable[CapturedLine]:
    output = run(*cmd, check=False)
    for line in output.stdout.strip().splitlines():
        with contextlib.suppress(AttributeError):
            yield CapturedLine(**regex_to_use.match(line).groupdict())


def check_code_mistakes() -> Iterable[CapturedLine]:
    """
    Yields a list of detected errors in the files.
    Falls back to flake8 if ruff doesn't exist on the system (which it should, as it's a dependency of this package!).
    """

    if has_executable("ruff"):
        yield from yield_from_regex("ruff", "check", "--no-fix", ".", regex_to_use=capturer)
    # elif has_executable("flake8"):
    #     yield from yield_from_regex("flake8", ".", regex_to_use=capturer)
    else:
        logger.warning("No code validation is done as ruff is not available.")


def check_formatting() -> Iterable[CapturedLine]:
    """
    Yields a list of files that would be reformatted.
    Falls back to black if ruff isn't there.
    """

    if has_executable("ruff"):
        yield from yield_from_regex("ruff", "format", "--check", ".", regex_to_use=formater)
    else:
        logger.warning("No format check is done because ruff is not available.")


src/ruff2bitbucket/bitbucket.py
"""
https://developer.atlassian.com/server/bitbucket/rest/v815/api-group-builds-and-deployments/#api-insights-latest-projects-projectkey-repos-repositoryslug-commits-commitid-reports-key-annotations-post
"""

import re
from functools import lru_cache
from urllib.parse import urlparse

from .git import get_current_git_commit_hash, get_current_repo_uri

__all__ = ["get_repo_info"]


class RepoInfo:
    def __init__(self) -> None:
        url = get_current_repo_uri()
        if not url:
            raise ValueError("No git has been set up, so we can't upload anything to BitBucket.")

        self.parse_result = urlparse(url)

        match = re.search(r"/scm/(?P<key>.*?)/(?P<slug>.*?)\.git", self.parse_result.path, re.IGNORECASE)
        if not match:
            match = re.search(
                r"/projects/(?P<key>.*?)/repos/(?P<slug>.*?)[/\\.].*$",
                self.parse_result.path,
                re.IGNORECASE,
            )

        if not match:
            raise ValueError(
                f"Couldn't interpret '{url}' as a BitBucket url. Please file an issue if this is in error!"
            )

        self.repo_key = match.group(1)
        self.repo_slug = match.group(2)
        self.commit_id = get_current_git_commit_hash()

    @property
    def report_endpoint(self) -> str:
        return (
            f"{self.parse_result.scheme}://{self.parse_result.netloc}/rest/insights/latest/projects/{self.repo_key}"
            f"/repos/{self.repo_slug}/commits/{self.commit_id}/reports/ruff2bitbucket"
        )

    @property
    def annotations_endpoint(self) -> str:
        return self.report_endpoint + "/annotations"

    @property
    def commit_url(self) -> str:
        return (
            f"{self.parse_result.scheme}://{self.parse_result.netloc}/projects/{self.repo_key}"
            f"/repos/{self.repo_slug}/commits/{self.commit_id}"
        )


@lru_cache(1)
def get_repo_info() -> RepoInfo:
    return RepoInfo()


src/ruff2bitbucket/credentials.py
from __future__ import annotations

import os
from argparse import ArgumentParser
from dataclasses import dataclass, field
from functools import lru_cache
from itertools import product
from typing import ClassVar, Iterator, List, Tuple

__all__ = ["get_credentials"]


@dataclass
class Credentials:
    correct_combinations: List[Tuple[str, str]] = field(default_factory=list)

    potential_user_envvars: ClassVar[Tuple[str, ...]] = ("USER", "USR")
    potential_pass_envvars: ClassVar[Tuple[str, ...]] = ("PASS", "PW", "PWD", "PASSWORD")

    _correct_combination: Tuple[str, str] | None = field(default=None, init=False)

    def __post_init__(self) -> None:
        user_vars = [
            (user_var, user_var[: -len(ending)])
            for user_var in os.environ
            for ending in self.potential_user_envvars
            if user_var.endswith(ending)
        ]
        pass_vars = [
            (pass_var, pass_var[: -len(ending)])
            for pass_var in os.environ
            for ending in self.potential_pass_envvars
            if pass_var.endswith(ending)
        ]

        for (user_var, user_prefix), (pass_var, pass_prefix) in product(user_vars, pass_vars):
            if user_prefix != pass_prefix:
                continue

            self.correct_combinations.append((os.getenv(user_var), os.getenv(pass_var)))

    def __len__(self) -> int:
        if self._correct_combination:
            return 1
        return len(self.correct_combinations)

    def __iter__(self) -> Iterator[Tuple[str, str]]:
        if self._correct_combination:
            yield self._correct_combination
        else:
            yield from self.correct_combinations

    def report_correct_combination(self, combo: Tuple[str, str]) -> None:
        self._correct_combination = combo


@lru_cache(1)
def get_credentials() -> Credentials:
    parser = ArgumentParser()

    parser.add_argument("--user", help="Specify a username", default=None, nargs="?")
    parser.add_argument("--pass", help="Specify a password", default=None, nargs="?")
    args = parser.parse_args()

    creds = Credentials()
    potential = (args.user, getattr(args, "pass"))  # I don't know how they do this... Isn't `pass` a reserved keyword??
    if all(potential):
        creds.report_correct_combination(potential)

    return creds


src/ruff2bitbucket/__main__.py
import json
import logging
import sys
from itertools import chain
from typing import List

import requests

from .bitbucket import get_repo_info
from .capturer import check_code_mistakes, check_formatting
from .common import CapturedLine
from .credentials import get_credentials

__all__ = ["main"]

logger = logging.getLogger(__name__)


def bitbucket_upload(upload_uri: str, report: dict, name: str, error_code: int) -> None:
    for credential in get_credentials():
        response = requests.put(upload_uri, json=report, auth=credential)
        if response.status_code == 401:  # Unauthorized
            continue

        get_credentials().report_correct_combination(credential)

        if response.status_code == error_code:
            logger.warning("'PUT %s' reported one or more errors:", upload_uri)
            text = json.dumps(response.json(), indent=4).splitlines()
            for line in text:
                logger.warning("%s", line)
        break  # But my authentication user was right, so we will stop here
    else:
        logger.error("Cannot upload the %s to bitbucket. No valid user/pass found.", name)
        sys.exit(1)


def upload_code_insights(upload_uri: str, captured_lines: List[CapturedLine]) -> None:
    annotations = []
    for message in captured_lines[:1_000]:  # Bitbucket can receive at most 1000 annotations
        annotations.append(
            {
                "reportKey": "ruff2bitbucket",
                "path": message.filename,
                "line": message.line,
                "message": message.description,
                "severity": "LOW",
                "type": "CODE_SMELL",
            }
        )

    bitbucket_upload(upload_uri, report={"annotations": annotations}, name="annotations", error_code=404)


def upload_code_statistics(upload_uri: str, captured_lines: List[CapturedLine]) -> None:
    need_reformat = sum(1 for cl in captured_lines if "reformat" in cl.description)

    report = {
        "result": "FAIL" if captured_lines else "PASS",
        "title": "ruff report",
        "reporter": "ruff2bitbucket",
        "report_type": "CODE_SMELL",
        "data": [
            {
                "title": "Need reformat",
                "type": "NUMBER",
                "value": need_reformat,
            },
            {
                "title": "Issue count",
                "type": "NUMBER",
                "value": len(captured_lines) - need_reformat,
            },
        ],
    }

    bitbucket_upload(upload_uri, report=report, name="report", error_code=400)


def main() -> None:
    logging.basicConfig()

    creds = get_credentials()
    if not creds:
        logger.error("No valid credentials found.")
        sys.exit(1)

    captured_lines = list(chain(check_code_mistakes(), check_formatting()))

    if not captured_lines:
        logger.info("no errors detected. No report will be uploaded.")
        sys.exit(0)

    upload_code_statistics(get_repo_info().report_endpoint, captured_lines)
    upload_code_insights(get_repo_info().annotations_endpoint, captured_lines)

    logger.info("reports were succesfully uploaded:\n%s", get_repo_info().commit_url)
